python/observation-publisher/publisher/prediction_log_consumer.py

```diff
@@ -0,0 +1,137 @@
+    KafkaConsumerConfig,
+    ObservationSource,
+    ObservationSourceConfig,
+)
+from publisher.observability_backend import ObservationSink
+from publisher.prediction_log_parser import (
+    PREDICTION_LOG_TIMESTAMP_COLUMN,
+    parse_struct_to_feature_table,
+    parse_struct_to_result_table,
+)
+
+
+class PredictionLogConsumer(abc.ABC):
+    @abc.abstractmethod
+    def poll_new_logs(self) -> List[PredictionLog]:
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def commit(self):
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def close(self):
+        raise NotImplementedError
+
+    def start_polling(
+        self, observation_sink: ObservationSink, inference_schema: InferenceSchema
+    ):
+        try:
+            while True:
+                logs = self.poll_new_logs()
+                if len(logs) == 0:
+                    continue
+                df = log_batch_to_dataframe(logs, inference_schema)
+                observation_sink.write(dataframe=df)
+                self.commit()
+        finally:
+            self.close()
+
+
+class KafkaPredictionLogConsumer(PredictionLogConsumer):
+    def __init__(self, config: KafkaConsumerConfig):
+        consumer_config = {
+            "bootstrap.servers": config.bootstrap_servers,
+            "group.id": config.group_id,
+        }
+
+        if config.additional_consumer_config is not None:
+            consumer_config.update(config.additional_consumer_config)
+
+        self._consumer = Consumer(consumer_config)
+        self._batch_size = config.batch_size
+        self._consumer.subscribe([config.topic])
+        self._poll_timeout = config.poll_timeout_seconds
+
+    def poll_new_logs(self) -> List[PredictionLog]:
+        messages = self._consumer.consume(self._batch_size, timeout=self._poll_timeout)
+        errors = [msg.error() for msg in messages if msg.error() is not None]
+
+        if len(errors) > 0:
+            print(f"Last encountered error: {errors[-1]}")
+            raise KafkaException(errors[-1])
+
+        return [
+            parse_message_to_prediction_log(msg.value().decode("utf-8"))
+            for msg in messages
+            if (msg is not None and msg.error() is None)
+        ]
+
+    def commit(self):
+        self._consumer.commit()
+
+    def close(self):
+        self._consumer.close()
+
+
+def new_consumer(config: ObservationSourceConfig) -> PredictionLogConsumer:
+    if config.type == ObservationSource.KAFKA:
+        return KafkaPredictionLogConsumer(config.kafka_config)
+    else:
+        raise ValueError(f"Unknown consumer type: {config.type}")
+
+
+def parse_message_to_prediction_log(msg: str) -> PredictionLog:
+    log = PredictionLog()
+    log.ParseFromString(msg)
+    return log
+
+
+def log_to_records(
+    log: PredictionLog, inference_schema: InferenceSchema
+) -> Tuple[List[List[np.int64 | np.float64 | np.bool_ | np.str_]], List[str]]:
+    request_timestamp = log.request_timestamp.ToDatetime()
```

Comment:
is this supported in arize? should we use the timestamp in second instead? based on this https://docs.arize.com/arize/sending-data-guides/model-schema-reference#6.-timestamp it expect integer unix second

Reply:
https://github.com/Arize-ai/client_python/blob/main/arize/pandas/validation/validator.py#L1013

Timestamp is accepted too. I have added a test case for validation.